src/Auth/Guards/KeycloakGuard.php

```diff
@@ -0,0 +1,228 @@
+use KeycloakAuthGuard\Auth\JwtPayloadUserProvider;
+use KeycloakAuthGuard\Exceptions\InvalidJwtTokenException;
+use KeycloakAuthGuard\Exceptions\UserNotFoundException;
+use KeycloakAuthGuard\JwtToken;
+use KeycloakAuthGuard\Services\RealmPublicKeyRetrieverInterface;
+use stdClass;
+
+class KeycloakGuard implements Guard
+{
+    private ?Authenticatable $user;
+    private ?stdClass $decodedToken;
+
+    /**
+     * @param RealmPublicKeyRetrieverInterface $publicKeyRetriever
+     * @param UserProvider $provider
+     * @param Request $request
+     */
+    public function __construct(
+        private readonly RealmPublicKeyRetrieverInterface $publicKeyRetriever,
+        private readonly UserProvider                     $provider,
+        private readonly Request                          $request
+    )
+    {
+        $this->user = null;
+        $this->decodedToken = null;
+        $this->authenticate();
+    }
+
+    /**
+     * Decode token, validate and authenticate user
+     *
+     * @return void
+     * @throws InvalidJwtTokenException
+     */
+    private function authenticate(): void
+    {
+        try {
+            $this->decodedToken = JwtToken::decode(
+                $this->getTokenForRequest(),
+                $this->publicKeyRetriever->getPublicKey(),
+                Config::get('keycloak.leeway')
+            );
+        } catch (Exception $e) {
+            throw new InvalidJwtTokenException($e->getMessage());
+        }
+
+        if ($this->decodedToken) {
+            $this->validate($this->getUserCredentials());
+        }
+    }
+
+    /**
+     * Get the token for the current request.
+     *
+     * @return string
+     */
+    private function getTokenForRequest(): string
+    {
+        if (!empty($this->request->bearerToken())) {
+            return $this->request->bearerToken();
+        }
+
+        if (!empty(Config::get('keycloak.input_key', ''))) {
+            return $this->request->input(
+                Config::get('keycloak.input_key', '')
+            );
+        }
+
+        return '';
+    }
+
+    /**
+     * Get the user credentials for retrieving of the user.
+     *
+     * @return array
+     */
+    private function getUserCredentials(): array
+    {
+        if ($this->provider instanceof JwtPayloadUserProvider) {
+            return $this->getCustomClaimsTokenData();
+        }
+
+        return [
+            Config::get('keycloak.user_provider_credential') => $this->getTokenPayloadData(Config::get('keycloak.token_principal_attribute'))
+        ];
+    }
+
+    private function getCustomClaimsTokenData(string $key = null, mixed $default = null): mixed
+    {
+        $customClaimsKey = Config::get('keycloak.jwt_payload_custom_claims_attribute');
+        $customClaimsTokenData = $this->getTokenPayloadData($customClaimsKey, []);
+
+        if (empty($key)) {
+            return $customClaimsTokenData;
+        }
+
+        return Arr::get($customClaimsTokenData, $key, $default);
+    }
```

Comment:
I think it would make sense to pass the `$default` argument to `getTokenPayloadData` as the second parameter. Otherwise the method caller might use the `$default` parameter to something else, but still unexpectedly receive an empty array as the return value.

```suggestion
    private function getCustomClaimsTokenData(string $key = null, mixed $default = null): mixed
    {
        $customClaimsKey = Config::get('keycloak.jwt_payload_custom_claims_attribute');
        $customClaimsTokenData = $this->getTokenPayloadData($customClaimsKey, $default);

        if (empty($key)) {
            return $customClaimsTokenData;
        }

        return Arr::get($customClaimsTokenData, $key, $default);
    }
```

Reply:
Based on the API of this function the value of the `$default` param is the value that will be returned in case if no other value is presented in custom claims. `$customClaimsTokenData` is a container that has all custom claims, I think it's logical to use array as the default value for `$customClaimsTokenData` and not use the `$default` because it represents the value that will be returned from the function and it can have any type.

Reply:
Ah yes, that makes sense. I was probably confused. 😅